### ISSUE TITLE
Bug with snippets using `$`, as PHP ones

### ISSUE THREAD

**Issue by xeluxee** [L3MON4D3/LuaSnip#174]
``` json
  "for …": {
    "prefix": "for",
    "body": [
      "for (\\$${1:i}=${2:0}; \\$$1 < $3; \\$$1++) { ",
      "\t${0:# code...}",
      "}"
    ],
    "description": "For-loop"
  },
```

- If I jump to locations without editing iterator name (by default `i`) everything goes fine
- If I edit iterator name (parameter `$1`) extra dollars appears:
![luasnip-issue](https://user-images.githubusercontent.com/88047141/133849313-62274da5-3e70-451e-b077-df04293f8a14.png)

**Comment by L3MON4D3** [comment]
Cannot reproduce with 
```lua
ls.snippets = {all={ls.parser.parse_snippet("for", "for (\\$${1:i}=${2:0}; \\$$1 < $3; \\$$1++) { \n\t${0:# code...}\n}")}}
```

https://user-images.githubusercontent.com/41961280/133851520-13161b5b-e059-491e-961b-69d2550a77c1.mp4

Could you try disabling other plugins/try the snippet in a buffer with another filetype until it works?

**Comment by xeluxee** [comment]
Just discovered that this happens when LSP client is attached. I'm using `intelephense`.

**Comment by L3MON4D3** [comment]
You could try disabling formatting (or other features, until it works) in `lspconfig.intelephense.setup{}`.

**Comment by xeluxee** [comment]
I wasn't able to find such a setting and I think formatting isn't related, because code isn't formatted when I expand a snippet

With vsnip this problem disappears

**Comment by L3MON4D3** [comment]
I found the dotfiles of someone who also opened an issue about formatting with intelephense, maybe try their options [link](https://github.com/knowler/dotfiles/blob/a08380648b97508845eddbc590e32b411500278a/.config/nvim/lua/lsp-config.lua#L86-L108)

**Comment by L3MON4D3** [comment]
The problem disappearing with vsnip is because they manually track changes done to text whereas luasnip uses extmarks.
Problem is, if a region of text that contains extmarks is changed via nvims api, the extmarks are pushed to the end, even if there wasn't any (visible) change, therefore autoformatting can mess luasnip up pretty bad

**Comment by xeluxee** [comment]
> I found the dotfiles of someone who also opened an issue about formatting with intelephense, maybe try their options [link](https://github.com/knowler/dotfiles/blob/a08380648b97508845eddbc590e32b411500278a/.config/nvim/lua/lsp-config.lua#L86-L108)

I tried this config but it doesn't seem to work :cry:

**Comment by L3MON4D3** [comment]
Dammit :/
Could you post the output of `lua Luasnip_current_nodes[vim.api.nvim_get_current_buf()].parent:dump()` after changing the text in `$1`?

**Comment by L3MON4D3** [comment]
Yep, the problem is that the second node ($1) is in the range `(2,5), (2,9)`, whereas it should start at `(2,6)` (going after the text in your first screenshots).

You could try asking around on gitter or matrix for how to disable formatting (maybe disabling it in the capabilities passed in setup could work, as a last resort)

**Comment by xeluxee** [comment]
Ok thanks, I'll investigate
